spot-electron/src/bt-beacon/daemon/darwin/bluetoothd.js

```diff
@@ -0,0 +1,159 @@
+
+const DEVICE_STATES = [ 'Unknown', 'Resetting', 'Unsupported', 'Unauthorized', 'Powered Off', 'Powered On' ];
+
+/**
+ * Implements a platform specific daemon to interact with the BLE hardware in the computer/device.
+ */
+class BluetoothD {
+    advertisingStatusUpdatedPromise;
+
+    /**
+     * Instantiates a new instance.
+     */
+    constructor() {
+        this.xpcConnection = new XpcConnection('com.apple.bluetoothd');
+
+        this.xpcConnection.on('error', message => {
+            logger.error('XpcError', message);
+        });
+
+        this.xpcConnection.on('event', event => {
+            const messageArguments = event.kCBMsgArgs;
+
+            switch (event.kCBMsgId) {
+            case 4:
+                // Device status update
+                this.setStatus(messageArguments.kCBMsgArgState);
+                break;
+            case 28:
+                // Beacon activation result
+                this.setAdvertisingStatus(!messageArguments.kCBMsgArgResult);
+                break;
+            case 29:
+                // Beacon deactivation result
+                this.setAdvertisingStatus(Boolean(messageArguments.kCBMsgArgResult));
+                break;
+            case 38:
+                // Device ready
+                this.readyPromise && this.readyPromise();
+                break;
+            case 67:
+                // Connection request. We don't need this.
+                break;
+            default:
+                logger.warn('Unknown XpcEvent', event);
+            }
+        });
+    }
+
+    /**
+     * Connects to the BT device (opens XPC channel).
+     *
+     * @returns {Promise<void>}
+     */
+    connect() {
+        return new Promise(resolve => {
+            this.readyPromise = resolve;
+            this.xpcConnection.setup();
+            this.sendMessage(1, {
+                kCBMsgArgName: `jitsi-bt-beacon-${(new Date()).getTime()}`,
+                kCBMsgArgOptions: {
+                    kCBInitOptionShowPowerAlert: 1
+                },
+                kCBMsgArgType: 1,
+                kCBMsgArgVersion: 20161219
+            });
+        });
+    }
+
+    /**
+     * Sends a message to the BT device.
+     *
+     * @param {number} id - The message ID to send.
+     * @param {Object} args - The message args to send.
+     * @returns {void}
+     */
+    sendMessage(id, args) {
+        const message = {
+            kCBMsgId: id
+        };
+
+        if (args) {
+            message.kCBMsgArgs = args;
+        }
+        this.xpcConnection.sendMessage(message);
+    }
+
+    /**
+     * Updates the advertising status of the device.
+     *
+     * @param {boolean} status - The status to set.
+     * @returns {void}
+     */
+    setAdvertisingStatus(status) {
+        this.advertisingStatus = status;
+        logger.info('BluetoothD advertising', this.advertisingStatus ? 'started' : 'stopped');
+        if (typeof this.advertisingStatusUpdatedPromise === 'function') {
```

Comment:
When would a promise be a typeof function?

Reply:
This is not a promise, this is the resolve function. Maybe needs a better naming?

Reply:
Yes, I would suggest renaming. Maybe denote it's a callback. Although I wonder then why have promises and callbacks.

Reply:
Renamed